// @foxglove/extension/src/experimental.ts
/* eslint-disable @typescript-eslint/no-namespace */
import type { Immutable } from "./immutable";
import type {
  ExtensionContext as BaseExtensionContext,
  PanelExtensionContext as BasePanelExtensionContext,
} from "./stable";

/**
 * The experimental namespace contains experimental APIs that are not yet stable and WILL change or
 * be REMOVED at any time.
 *
 * These APIS are provided for isolated testing and feedback.You should avoid shipping org
 * extensions which depend on them because they WILL break in different versions of Foxglove and
 * can be CHANGED or REMOVED at any time.
 *
 * @hidden
 */
export namespace Experimental {
  /**
   * This type represents the arguments you pass to {@link ExtensionContext.registerDataLoader}
   * when you want to register a data loader.
   *
   * @category Data loaders
   */
  export type RegisterLocalFileDataLoaderArgs = {
    /** The type of data loader. Must be "file" for local file data loaders. */
    type: "file";
    /** The URL to the WebAssembly module for this data loader. */
    wasmUrl: string;
    /** The file extension that this data loader should be associated with. */
    supportedFileType: `.${string}`;
    /** Whether the data loader supports being passed multiple files */
    supportsMultiFile?: boolean;
  };

  /**
   * Arguments for registering a custom data loader. Currently, only local file data loaders are
   * supported.
   *
   * See: {@link RegisterLocalFileDataLoaderArgs}
   *
   * @category Data loaders
   */
  export type RegisterDataLoaderArgs = RegisterLocalFileDataLoaderArgs;

  /**
   * This type represents the arguments you pass to {@link ExtensionContext.registerPanel}.
   *
   * @category Custom panels
   */
  export type ExtensionPanelRegistration = {
    /**
     * Unique name of the panel within your extension
     *
     * NOTE: Panel names within your extension must be unique. The panel name identifies this panel
     * within a layout. Changing the panel name will cause layouts using the old name unable to load
     * your panel.
     */
    name: string;

    /**
     * This function is invoked when your panel is initialized
     *
     * @returns (optional) A function which will be called when the panel is removed or replaced.
     * Perform any cleanup logic here to gracefully tear down your panel.
     */
    initPanel: (context: PanelExtensionContext) => void | (() => void);
  };

  /**
   * See: {@link ExtensionContext.registerUserScriptUtility}
   */
  export type RegisterUserScriptUtilityArgs = {
    name: `${string}.ts`;
    source: string;
  };

  /**
   * The arguments you pass to {@link ExtensionContext.registerVideoDecoder}
   * when you want to register a custom video decoder.
   *
   * @category Video decoders
   */
  export type RegisterVideoDecoderArgs = {
    /**
     * Builder function that creates decoder instances, provided as a string. The function receives
     * callbacks for output frames and errors, and returns a decoder backend instance.
     *
     * See {@link VideoFrameDecoderBuilder} for the function signature of the builder function,
     *
     * IMPORTANT: The factory must be serializable (no imports or external closures). Extensions
     * must embed or otherwise manage any external resources (e.g., WASM or script assets) required
     * by the decoder.
     */
    decoderBuilderSource: string;

    /** Human-readable name for display in the app interface. */
    displayName: string;

    /**
     * Unique identifier for this decoder.
     *
     * Must be stable across extension versions and unique across all extensions. Recommended
     * format: "org.example.decoder-name" or similar namespaced format.
     */
    id: string;
  };

  /**
   * Abstract video decoder interface that both native VideoDecoder and custom decoders can
   * implement.
   *
   * @category Video decoders
   */
  export type VideoFrameDecoder = Pick<
    VideoDecoder,
    "close" | "configure" | "decode" | "reset" | "state" | "decodeQueueSize" | "ondequeue"
  >;

  /**
   * Factory function that creates decoder instances
   *
   * @category Video decoders
   */
  export type VideoFrameDecoderBuilder = (init: VideoDecoderInit) => VideoFrameDecoder;

  export interface ExtensionContext extends BaseExtensionContext {
    /**
     * Register a data loader to add support for additional file formats.
     *
     * A data loader teaches Foxglove how to read new file formats. The data loader is a WASM layer
     * that bridges the semantics of the custom file format to Foxglove.
     *
     * See: {@link RegisterDataLoaderArgs}
     */
    registerDataLoader(args: RegisterDataLoaderArgs): void;

    /**
     * `registerPanel` adds a new panel to the Foxglove interface. To register a panel you provide a
     * `name` and an `initPanel` function.
     *
     * The `initPanel` function accepts a {@link PanelExtensionContext} argument, which contains
     * properties and methods for accessing panel data and rendering UI updates. It also returns an
     * optional cleanup function to run when the extension `panelElement` unmounts.
     */
    registerPanel(params: ExtensionPanelRegistration): void;

    /**
     * Register a new user-script utility.
     *
     * This utility is available to all user-scripts and imported by its name.
     *
     * Example registration:
     *
     * ```
     * registerUserScriptUtility({
     *   name: "my_add.ts",
     *   source: `
     * export function add(a: number, b: number) {
     *  return a + b;
     * }
     * `);
     * ```
     *
     * You can also keep the source code in a file and import it like this:
     *
     * ```
     * import my_add from "./my_add.ts?raw";
     *
     * registerUserScriptUtility({ name: "my_add.ts", source: my_add });
     * ```
     *
     * Example usage in a user-script:
     *
     * ```
     * import { add } from "./my_add.ts";
     *
     * // ... call `add` in your script function or wherever needed
     * ```
     *
     * Note: The filename must contain only alphanumeric characters, dashes, underscores, and end with .ts (e.g., "my_utility.ts")
     *
     * @param name
     * @param source
     */
    registerUserScriptUtility(args: RegisterUserScriptUtilityArgs): void;

    /**
     * Register a custom video decoder backend.
     *
     * This allows extensions to provide custom video decoders for codecs that may not be supported
     * by the native WebCodecs API.
     */
    registerVideoDecoder(args: RegisterVideoDecoderArgs): void;
  }

  export interface ExtensionModule {
    /**
     * This function will be called when your extension is loaded. In this function, you can register
     * your custom panels or other types of extension features.
     *
     * The function may return a Promise if async initialization is needed before registering
     * extension features. The extension will not be considered fully activated until the Promise
     * resolves.
     */
    activate: ExtensionActivate;
  }

  export type ExtensionActivate = (extensionContext: ExtensionContext) => void | Promise<void>;

  /**
   * A button in the panel toolbar.
   */
  export type PanelToolbarButton = {
    type: "button";
    /** Unique identifier for the button. */
    id: string;
    /**
     * The icon to display in the button. It can be a predefined icon name from {@link PanelToolbarPredefinedIcon} or a SVGElement.
     * If passing a SVGElement, it should be 16x16px and use the `currentColor` attribute for variants to work correctly.
     * */
    icon: SVGElement;
    /** Tooltip text to display when hovering over the button. */
    tooltip?: string;
    /** Function to call when the button is clicked. */
    onClick?: (event: MouseEvent) => void;
    /** Whether the button is disabled. */
    disabled?: boolean;
    /** Data test id to apply to the button, useful for testing. */
    dataTestId?: string;
    /** Variant of the button. */
    variant?: "regular" | "solid";
  };

  export type PanelToolbarLink = {
    type: "link";
    /** Unique identifier for the link. */
    id: string;
    /**
     * The icon to display in the button. The SVGElement should be 16x16px and use the `currentColor`
     * attribute for variants to work correctly.
     * */
    icon: SVGElement;
    /** Tooltip text to display when hovering over the button. */
    tooltip?: string;
    /** The URL to navigate to when the link is clicked. */
    href: string;
    /** Whether the link is disabled. */
    disabled?: boolean;
    /** Data test id to apply to the button, useful for testing. */
    dataTestId?: string;
    /** Variant of the button. */
    variant?: "regular" | "solid";
    /** Target of the link. Defaults to `_blank`. */
    target?: "_blank" | "_self";
  };

  /**
   * A vertical divider in the panel toolbar used to separate items.
   */
  export type PanelToolbarDivider = {
    type: "divider";
    /** Unique identifier for the divider. */
    id: string;
  };

  /**
   * Items that can be displayed in the panel toolbar.
   */
  export type PanelToolbarItem = PanelToolbarButton | PanelToolbarLink | PanelToolbarDivider;

  export interface PanelExtensionContext extends BasePanelExtensionContext {
    /**
     * Set the items in the panel toolbar.
     */
    setPanelToolbarItems: (items: Immutable<PanelToolbarItem[]>) => void;
  }
}
